### PR TITLE
refactor: Waku Cache & New Architecture

### PR DESCRIPTION
**Pull request by SionoiS**
Opening this to get feedback on a new way to architecture Nwaku.

This serve just as an example. I use Chronos `AsyncEventQueue` as some kind of broadcast channel to send the information needed to various sub-systems.

Relay messages are sent to Waku Cache and also topic subscriptions they are then filtered and sent to Discv5 and Relay.
It allow for a slimmer `waku_node` too since `waku_node` no longer be called just to then call the inner protocols.

Please focus your attention on the information flow between; Discv5, Relay, Filter, LightPush and the APIs

Waku Cache could be removed if the topic filtering is duplicated in Relay and Discv5. the caches could also live somewhere else. 

WDYT?

---

P.S.

[Nimbus usage of AsyncEventQueue](https://github.com/search?q=repo%3Astatus-im%2Fnimbus-eth2%20AsyncEventQueue&type=code)

I also keep [notes](https://www.notion.so/S-P-s-Architectural-Notes-481dec3dc54d41ddba9f45e440f85bca) about the architecture of Nwaku. Feel free to add to it your wishes or observations.

### CODE REVIEW

**Comment by SionoiS** [comment]
Tagging @NagyZoltanPeter @jm-clius @Ivansete-status

**Review by Ivansete-status — CHANGES_REQUESTED** [review]
Good morning! Thanks for this initiative!

Few points we need to address before

- I started the review but then I realized that the changes don't compile
- `AsyncEventQueue` seems not to be from the standard library. We need to ask Jacek or Zahary about introducing this concept in Waku, in case they see any possible drawback or flaw.
- I can't see where the messages are emitted in general. To me, it is difficult to follow the flow.

File: waku/waku_cache.nim [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+
+type
+  WakuCache* = ref object
+    # Input
+    rawSubscriptionQueue: AsyncEventQueue[SubscriptionEvent]
+    messageQueue: AsyncEventQueue[(PubsubTopic, WakuMessage)]
+
+    # State
+    started: bool
+    subscriptions: Table[PubsubTopic, tuple[manual: bool, count: int]]
```
[the remark this answers is not in the document]
Comment (SionoiS) [reply waku/waku_cache.nim:38]: Manual subscription to a shard AKA subcribe to a pubsub topic that is a shard define for the Waku Network.

File: waku/waku_cache.nim [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+      continue
+
+    wc.subscriptions.mgetOrPut(topic, (false, 0)).manual = false
+
+proc batchContentSubscribe(wc: WakuCache, batch: seq[ContentTopic]) =
+  for topic in batch:
+    if wc.messageCache.isSubscribed(topic):
+      continue
+
+    wc.messageCache.subscribe(topic)
```
[the remark this answers is not in the document]
Comment (SionoiS) [reply waku/waku_cache.nim:92]: Yes exactly

File: waku/waku_archive/archive.nim [hunk 3]
```diff
@@ -118,6 +126,18 @@ proc handleMessage*(w: WakuArchive,
   let insertDuration = getTime().toUnixFloat() - insertStartTime
   waku_archive_insert_duration_seconds.observe(insertDuration)
 
+proc wakuMessageListener(wa: WakuArchive) {.async.} =
+  let key = wa.messageQueue.register()
+
+  while wa.started:
+    let events = await wa.messageQueue.waitEvents(key)
```
[the remark this answers is not in the document]
Comment (SionoiS) [reply waku/waku_archive/archive.nim:133]: They are emitted in relay, but the goal is exactly to not care where they are emitted from. They could also come from light push.

File: waku/waku_archive/archive.nim [hunk 4]
```diff
@@ -118,6 +126,18 @@ proc handleMessage*(w: WakuArchive,
   let insertDuration = getTime().toUnixFloat() - insertStartTime
   waku_archive_insert_duration_seconds.observe(insertDuration)
 
+proc wakuMessageListener(wa: WakuArchive) {.async.} =
+  let key = wa.messageQueue.register()
```
[the remark this answers is not in the document]
Comment (SionoiS) [reply waku/waku_archive/archive.nim:130]: I'll see what can be done

File: waku/waku_archive/archive.nim [hunk 5]
```diff
@@ -241,3 +261,10 @@ proc startMessageRetentionPolicyPeriodicTask*(w: WakuArchive,
     discard setTimer(Moment.fromNow(interval), executeRetentionPolicy)
 
   discard setTimer(Moment.fromNow(interval), executeRetentionPolicy)
+
+proc start*(wa: WakuArchive) {.async.} =
+  wa.started = true
+
+  asyncSpawn wa.wakuMessageListener()
+
+proc stop*(wa: WakuArchive) {.async.} = wa.started = false
```
[the remark this answers is not in the document]
Comment (SionoiS) [reply waku/waku_archive/archive.nim:270]: :thinking: that's a good question there's probably a way to cancel futures no?

For archive, we could tie the periodic policy with the new lifecycle too.

**Comment by SionoiS** [comment]
> * I started the review but then I realized that the changes don't compile

Yes just wanted to ask for feedback as the changes would be major.

> * `AsyncEventQueue` seems not to be from the standard library. We need to ask Jacek or Zahary about introducing this concept in Waku, in case they see any possible drawback or flaw.

I'm in contract with @cheatfate who authored `AsyncEventQueue` I asked for more details.

> * I can't see where the messages are emitted in general. To me, it is difficult to follow the flow.

Might be because the setup phase is missing. In `app` the various sub-systems would be initialized with the queues.

File: waku/waku_archive/archive.nim [hunk 6]
```diff
@@ -72,22 +72,30 @@ type
...
-                    retentionPolicy.get()
-                  else:
-                    nil
-
-  let wakuArch = WakuArchive(driver: driver,
-                             validator: DefaultMessageValidator(),
-                             retentionPolicy: retPolicy)
+  driver: ArchiveDriver,
+  retentionPolicy = none(RetentionPolicy),
+  messageQueue: AsyncEventQueue[(PubsubTopic, WakuMessage)],
```
[the remark this answers is not in the document]
Comment (SionoiS) [reply waku/waku_archive/archive.nim:81]: My mistake

**Review by jm-clius — COMMENTED** [review]
Without reviewing the implementation details here, I think this is an architecture worth exploring provided we understand the `AsyncEventQueue` and impact. Since the specific problem you were trying to solve is simply related to listening to subscription changes, why not drop the `messageQueue` for now and only explore the `subscriptionQueue`? I'm also not sure about having everything into a single "WakuCache" - in future, perhaps just have different caches for subscriptions and messages with no need to combine them into a single structure?

**Comment by SionoiS** [comment]
> Without reviewing the implementation details here, I think this is an architecture worth exploring provided we understand the `AsyncEventQueue` and impact. Since the specific problem you were trying to solve is simply related to listening to subscription changes, why not drop the `messageQueue` for now and only explore the `subscriptionQueue`?

Sounds reasonable.

> I'm also not sure about having everything into a single "WakuCache" - in future, perhaps just have different caches for subscriptions and messages with no need to combine them into a single structure?

It made sense at the time but yes it could be separate.

**Comment by SionoiS** [comment]
> Since the specific problem you were trying to solve is simply related to listening to subscription changes, why not drop the `messageQueue` for now and only explore the `subscriptionQueue`?

Kinda hard to mix both.

> I'm also not sure about having everything into a single "WakuCache" - in future, perhaps just have different caches for subscriptions and messages with no need to combine them into a single structure?

Made sense to do it this way but it can be separate.

**Comment by SionoiS** [comment]
I can still see many advantages to this but since we don't have `AsyncChannel[T]` yet, Nimbus team have a lot on their plate and I  haven't though about how to incrementally do it, I'm putting this on pause for now.

**Comment by Ivansete-status** [comment]
> I gained knowledge about Nwaku in the process, that's a win for me, even if we never revisit this.

Yes indeed! That's a win for all of us of course! We can revisit this again once the `AsyncChannel[T]` gets more mature

**Pull request closed without merging**